src/client/eos-client.ts

```diff
@@ -37,30 +38,79 @@ export class EOSClient {
     return this.client.connect(onMessage)
   }
 
+  public send<T extends OutboundMessage<{}>>(
+    type: OutboundMessageType,
+    messageOptions: T,
+    ...messageTypes: InboundMessageType[]
+  ) {
+    return this.sendAndListen<T>(
+      messageOptions,
+      messageOptions.req_id === undefined ? type : messageOptions.req_id,
+      messageOptions.listen === undefined ? true : messageOptions.listen,
+      ...messageTypes
+    )
+  }
+
   public getActions(parameters: GetActionsMessageParameters, options: StreamOptions = {}) {
     options = withDefaults({ listen: true }, options)
-    this.client.send(getActionsMessage(parameters, options))
-
-    return this.createListener(options.requestId!, InboundMessageType.ACTION_TRACE)
+    const messageOptions = getActionsMessage(parameters, options)
+
+    return this.sendAndListen<OutboundMessage<GetActionsMessageBackendParameters>>(
+      messageOptions,
+      options.requestId!,
+      options.listen!,
+      InboundMessageType.ACTION_TRACE
+    )
   }
 
   public getTableRows(parameters: GetTableRowsMessageParameters, options: StreamOptions = {}) {
     options = withDefaults({ listen: true }, options)
-    this.client.send(getTableRowsMessage(parameters, options))
-
-    return this.createListener(options.requestId!, InboundMessageType.TABLE_DELTA)
+    const messageOptions = getTableRowsMessage(parameters, options)
+    return this.sendAndListen<OutboundMessage<GetTableRowsMessageBackendParameters>>(
+      messageOptions,
+      options.requestId!,
+      options.listen!,
+      InboundMessageType.TABLE_DELTA
+    )
   }
 
   public getTransaction(id: string, options: StreamOptions = {}) {
     options = withDefaults({ fetch: true, listen: true }, options)
-    this.client.send(getTransactionMessage(id, options))
+    const messageOptions = getTransactionMessage(id, options)
+
+    return this.sendAndListen<OutboundMessage<{ id: string }>>(
+      messageOptions,
+      options.requestId!,
+      options.listen!,
+      InboundMessageType.TRANSACTION_LIFECYCLE
+    )
+  }
+
+  private sendAndListen<T extends OutboundMessage<{}>>(
+    messageOptions: T,
+    requestId: string,
+    listen: boolean,
+    ...messageTypes: InboundMessageType[]
+  ) {
+    if (listen) {
+      return this.createListenerWithSend<T>(requestId!, messageOptions, ...messageTypes)
+    }
 
-    return this.createListener(options.requestId!, InboundMessageType.TRANSACTION_LIFECYCLE)
+    this.client.send(messageOptions)
+    return null
   }
 
-  private createListener(requestId: string, ...handledMessageTypes: InboundMessageType[]) {
+  private createListenerWithSend<T extends OutboundMessage<{}>>(
+    requestId: string,
+    parameters: T,
+    ...messageTypes: InboundMessageType[]
+  ) {
     const listen = (callback: ClientMessageListener) => {
-      this.registerListener({ handledMessageTypes, requestId, callback })
+      try {
```

Comment:
No need for a `try/finally` clause here, everything will be linear, "equivalent" to:

    this.registerListener({ messageTypes, requestId, callback })
    this.client.send(parameters)

But not quite, because `registerListener` throws somehow, then `send` will be called, which I think is not "right"